reference/security: add encryption-at-rest doc (#2410)
pingcap/docs: reference/security/encryption-at-rest.md

TiKV supports encryption-at-rest starting from v4.0.0. The feature allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation)  mode. To enable encryption-at-rest, an encryption key must be provided by user and this key is called master key. The suggested way to provide the master key is via AWS KMS, but specifying a key in plaintext stored in a file is also supported. TiKV automatically rotate data keys that it used to encrypt actual data files. Manually rotating the master key can be done occassionally. Note that encryption-at-rest only encrypts data at rest (i.e. on disk) and not while data is transferred over network. It is advices to use TLS together with encryption-at-rest.

Also from v4.0.0, BR supports S3 server-side encryption (SSE) when backup to S3. A customer owned AWS KMS key can also be used together with S3 server-side encrytion.

> ## TiKV Encryption-At-Rest

Author: TODO: add a section for Grafana monitoring for encryption.

region = "us-west-2"
```

### Caveat

The current version of TiKV encryption has some drawbacks that need to be noted when use. We are working actively to address those issues and improvments are in the pipe for future versions.

* When deploying a TiDB cluster, the majority of user data are stored in TiKV nodes, and those data are encrypted-at-rest when encryption is enabled. However there are small amount of user data stored in PD nodes as metadata. As of v4.0.0, PD doesn't support encryption-at-rest. As a workaround, it is recommended to use storage level encryption (e.g. file system encryption) to protect sensitive data stored in PD.
* TiKV currently does not exclude encryption keys and user data from core dump. It is highly adviced to disable core dump for TiKV process when using encryption-at-rest. This is not currently handled by TiKV itself.
* TiKV keeps track of encryption key and method used for each data files, using absolute path of the files. As a result, once encryption is turned on for a TiKV node, user should not change data file paths config such as `storage.data-dir`, `raftstore.raftdb-path`, `rocksdb.wal-dir` and `raftdb.wal-dir`.

> ## BR S3 server-side encryption

Author: TODO: copy this section to BR document.

Author: @overvenus do you think this is needed, and if so, which file should I modify?

---
title: Encryption-At-Rest
summary: Learn how to enable encryption-at-rest to protect sensitive data.
category: reference
---

# Encryption-At-Rest for TiKV <span class="version-mark">New in v4.0.0</span>

> Encryption-at-rest means that data is encrypted when it is stored. As opposed to encryption in flight (TLS) or encryption in use (rarely used). Different things could be doing encryption-at-rest (SSD drive, file system, cloud vendor, etc), but if the database does encryption by itself this helps ensure that attackers must authenticate with the database to gain access to data. For example, even when an attacker gains access to the physical machine, he cannot access data by copying files on disk.

Reviewer: ```suggestion
Encryption-at-rest means that data is encrypted when it is stored. For databases, this feature is also referred to as TDE (transparent data encryption). This is opposed to encryption in flight (TLS) or encryption in use (rarely used). Different things could be doing encryption-at-rest (SSD drive, file system, cloud vendor, etc), but by having TiKV do the encryption before storage this helps ensure that attackers must authenticate with the database to gain access to data. For example, when an attacker gains access to the physical machine, data cannot be accessed by copying files on disk.
```

> TiKV supports encryption-at-rest starting from v4.0.0. The feature allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation)  mode. To enable encryption-at-rest, an encryption key must be provided by user and this key is called master key. The suggested way to provide the master key is via AWS KMS, but specifying a key in plaintext stored in a file is also supported. TiKV automatically rotate data keys that it used to encrypt actual data files. Manually rotating the master key can be done occassionally. Note that encryption-at-rest only encrypts data at rest (i.e. on disk) and not while data is transferred over network. It is advices to use TLS together with encryption-at-rest.

Reviewer: ```suggestion
TiKV supports encryption-at-rest starting from v4.0.0. The feature allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation) mode. To enable encryption-at-rest, an encryption key must be provided by user and this key is called master key. The master key can be provided via AWS KMS (recommended), or specifying a key stored as plaintext in a file. TiKV automatically rotates data keys that it used to encrypt actual data files. Manually rotating the master key can be done occasionally. Note that encryption-at-rest only encrypts data at rest (i.e. on disk) and not while data is transferred over network. It is advised to use TLS together with encryption-at-rest.
```

> Also from v4.0.0, BR supports S3 server-side encryption (SSE) when backup to S3. A customer owned AWS KMS key can also be used together with S3 server-side encrytion.

Reviewer: ```suggestion
Also from v4.0.0, BR supports S3 server-side encryption (SSE) when backing up to S3. A customer owned AWS KMS key can also be used together with S3 server-side encrytion.
```

## TiKV Encryption-At-Rest

### Overview

> TiKV currently support encrypting data using AES128, AES192 or AES256, in CTR mode. There are two types of keys used in TiKV when encryption is enabled:

Reviewer: ```suggestion
TiKV currently support encrypting data using AES128, AES192 or AES256, in CTR mode. There are two types of keys used in TiKV when encryption is enabled, this is referred to as envelope encryption:
```

1. Master key. The master key is provided by user and is used to encrypt the data keys TiKV generates. Management of master key is external to TiKV.
2. Data key. The data key is generated by TiKV and is the key actually being used to encrypt data. The data key is automatically rotated by TiKV.

> The same master key can be shared by multiple instances of TiKV. The recommended way to provide a master key in production is via AWS KMS. User would need to create a CMK through AWS KMS, and then provide the CMK key id to TiKV in config file. The TiKV process would need access to the KMS CMK while it is running, which can be done by using [IAM](https://aws.amazon.com/iam/). If TiKV fail to get access to the KMS CMK, it will failed to start or restart. If TiKV loss access to the KMS CMK while it is running, data key rotation will be temporarily disabled. Please refer to AWS documentation for KMS and IAM usage.

Reviewer: ```suggestion
The same master key can be shared by multiple instances of TiKV. The recommended way to provide a master key in production is via AWS KMS. Create a CMK through AWS KMS, and then provide the CMK key id to TiKV in the config file. The TiKV process needs access to the KMS CMK while it is running, which can be done by using an [IAM role](https://aws.amazon.com/iam/). If TiKV fail to get access to the KMS CMK, it will fail to start or restart. If TiKV loses access to the KMS CMK while it is running, data key rotation will be temporarily disabled. Please refer to AWS documentation for KMS and IAM usage.
```

> Alternatively, if using custom key is desired, supplying master key via file is also supported. The file need to contain a 256 bits (or 32 bytes) key encoded as hex string. The file should end with a newline (i.e. "\n") and contain nothing else. Persistenting the key on disk, however, leaks the key, so the key file is only suitable to be stored on tempfs.

Reviewer: ```suggestion
Alternatively, if using custom key is desired, supplying the master key via file is also supported. The file needs to contain a 256 bits (or 32 bytes) key encoded as hex string. The file should end with a newline (i.e. "\n") and contain nothing else. Persisting the key on disk, however, leaks the key, so the key file is only suitable to be stored on tempfs in RAM.
```

Reviewer: Why do we require a newline? It seems like we could chomp new lines if they exist?

Author: I would like to make the file format very strict to make sure user aware of the required key length. I would like to do without the newline, but text editors (at least vim) tend to add a extra newline at line end, which is not a bad idea to indicate end of the key.

Reviewer: If I had to choose one, I would probably not have a newline. I generally agree that being strict about inputs is a much better approach, but chomping newlines is the one thin I don't see a problem with. I don't think it will cause any confusion about the file format and they can potentially avoid some deployment delays.

Author: we can make the newline optional, but it is too late for 4.0 GA branch cut.

Author: file key are not for production use anyway.

> Data keys are generated by TiKV and pass to underlying storage engine (i.e. RocksDB). All files written by RocksDB, including SST files, WAL files, and the MANIFEST file,  is encrypted by the current data key. Other temporary files used by TiKV that may include user data are also encrypted using the same data key. By default, data keys are automatically rotated by TiKV every week, for which the period is configurable. On key rotation, TiKV does not rewrite all existing files to replace the key, but RocksDB compaction are expected to rewrite old data into new data files, with the most recent data key, if the cluster gets constant write workload. TiKV keeps track of the key and encryption method used to encrypt each of of files and use the information to decrypt the content on reads.

Reviewer: ```suggestion
Data keys are generated by TiKV and passed to the underlying storage engine (i.e. RocksDB). All files written by RocksDB, including SST files, WAL files, and the MANIFEST file, are encrypted by the current data key. Other temporary files used by TiKV that may include user data are also encrypted using the same data key. Data keys are automatically rotated by TiKV every week by default, but the period is configurable. On key rotation, TiKV does not rewrite all existing files to replace the key, but RocksDB compaction is expected to rewrite old data into new data files, with the most recent data key, if the cluster gets constant write workload. TiKV keeps track of the key and encryption method used to encrypt each of of files and uses the information to decrypt the content on read.
```

> Regardless of data encryption method, data keys are encrypted using AES256 in GCM mode for additional authentication. This is why when passing master key from file the key has to be of 256 bits.

Reviewer: ```suggestion
Regardless of data encryption method, data keys are encrypted using AES256 in GCM mode for additional authentication. This requires the master key to be 256 bits.
```

Author: For KMS the key length discrepancy is handed by another layer of envelope encryption, so it doesn't matter.

### Configuring Encryption

To enable encryption, you can add the encryption section in TiKV's config file:

```
[security.encryption]
data-encryption-method = aes128-ctr
data-key-rotation-period = 7d
```

> Possible values for `data-encryption-method` are "aes128-ctr", "aes-192-ctr", "aes256-ctr" and "plaintext". The default value is "plaintext", which means encryption is not turned on. `data-key-rotation-period` defines how often TiKV rotate data key. Encryption can be turned on for a fresh TiKV cluster, or an existing TiKV cluster, though only data written after encryption being enabled are guaranteed to get encrypted. To disable encryption, remove `data-encryption-method` in config file, or reset it to "plaintext", and restart TiKV. To change encryption method, update `data-encryption-method` in config file and restart TiKV.

Reviewer: ```suggestion
Possible values for `data-encryption-method` are "aes128-ctr", "aes-192-ctr", "aes256-ctr" and "plaintext". The default value is "plaintext", which means encryption is not turned on. `data-key-rotation-period` defines how often TiKV rotates the data key. Encryption can be turned on for a fresh TiKV cluster, or an existing TiKV cluster, though only data written after encryption is enabled is guaranteed to be encrypted. To disable encryption, remove `data-encryption-method` in the config file, or reset it to "plaintext", and restart TiKV. To change encryption method, update `data-encryption-method` in config file and restart TiKV.
```

Reviewer: If I change it back to plaintext, at what point will all data be decrypted?

Author: along with rocksdb compaction, which is done lazily.

> Master key has to be specified if encryption is enabled (i.e. `data-encryption-method` is not "plaintext"). To specify a AWS KMS CMK as master key, add the `encryption.master-key` section after the `encryption` section:

Reviewer: ```suggestion
The master key has to be specified if encryption is enabled (i.e. `data-encryption-method` is not "plaintext"). To specify an AWS KMS CMK as master key, add the `encryption.master-key` section after the `encryption` section:
```

```
[security.encryption.master-key]
type = "kms"
key-id = "0987dcba-09fe-87dc-65ba-ab0987654321"
region = "us-west-2"
endpoint = "https://kms.us-west-2.amazonaws.com"
```

> The `key-id` specifies the key id for the KMS CMK. The `region` is the AWS region name for the KMS CMK. The `endpoint` is optional and don't need to be specified normally, unless you are using a AWS KMS compatible service from a non-AWS vendor.

Reviewer: ```suggestion
The `key-id` specifies the key id for the KMS CMK. The `region` is the AWS region name for the KMS CMK. The `endpoint` is optional and doesn't need to be specified normally, unless you are using an AWS KMS compatible service from a non-AWS vendor.
```

To specify a master key that's stored in a file, the master key config would look like the following:

```
[security.encryption.master-key]
type = "file"
path = "/path/to/key/file"
```

Here `path` is the path to the key file. The file must contain a 256 bits (or 16 bytes) key encoded as hex string, end with a newline ("\n") and contain nothing else. Example of the file content:

```
3b5896b5be691006e0f71c3040a29495ddcad20b14aff61806940ebd780d3c62
```

> ### Rotating Master Key

Reviewer: ```suggestion
### Rotating the Master Key
```

> To rotate master key, you have to specify both of the new master key and old master key in the config, and restart TiKV. Use `encryption.master-key` to specify the new master key, and use `encryption.previous-master-key` to specify the old master key. The config format for `encryption.previous-master-key` is the same as `encryption.master-key`. On restart TiKV would need access to both of the old and new master key, but once TiKV is up and running, TiKV will only need access to the new key. It is okay to leave the `encryption.previous-master-key` config in the config file from that on. Even on restart, TiKV will only try to use the old key if it fail to decrypt existing data using the new master key.

Reviewer: ```suggestion
To rotate the master key, you have to specify both the new master key and old master key in the config file, and restart TiKV. Use `encryption.master-key` to specify the new master key, and use `encryption.previous-master-key` to specify the old master key. The config format for `encryption.previous-master-key` is the same as `encryption.master-key`. On restart TiKV must access both the old and new master key, but once TiKV is up and running, TiKV will only need access to the new key. It is okay to leave the `encryption.previous-master-key` config in the config file from then on. Even on restart, TiKV will only try to use the old key if it fail to decrypt existing data using the new master key.
```

> Currently we don't support online master key rotation. Rotating master key would require restarting TiKV.

Reviewer: ```suggestion
Currently we don't support online master key rotation: a restart of TiKV is required.
```

Here is an example config for rotating the KMS CMK:

```
[security.encryption.master-key]
type = "kms"
key-id = "50a0c603-1c6f-11e6-bb9e-3fadde80ce75"
region = "us-west-2"

[security.encryption.previous-master-key]
type = "kms"
key-id = "0987dcba-09fe-87dc-65ba-ab0987654321"
region = "us-west-2"
```

### Monitoring

> To monitor encryption-at-rest, if you deploy TiKV with Grafana, you can look at the "Encryption" pannel in "TiKV-Details" dashboard. There are a few metrics to look for:

Reviewer: ```suggestion
To monitor encryption-at-rest, if you deploy TiKV with Grafana, you can look at the "Encryption" panel in the "TiKV-Details" dashboard. There are a few metrics to look for:
```

* Encryption initialized: 1 if encryption is initialized during TiKV startup, 0 otherwise. In case of master key rotation, after encryption is initialized, TiKV do not need access to the previous master key.
* Encryption data keys: number of existings data keys. The number is bumped by 1 after each time data key rotation happened. Use this metrics to monitor if data key rotation works as expected.
* Encrypted files: number of encrypted data files currently exists. Compare this number to existings data files in the data directory to estimate portion of data being encrypted, when turning on encryption for a previously unencrypted cluster.
* Encryption meta file size: size of the encryption meta data files.

> * Read/Write encryption meta duration: the extra overhead for operate on encryption metadata.

Reviewer: ```suggestion
* Read/Write encryption meta duration: the extra overhead to operate on encrypted metadata.
```

### Caveats

> The current version of TiKV encryption has some drawbacks that need to be noted when use. We are working actively to address those issues and improvments are in the pipe for future versions.

Reviewer: ```suggestion
The current version of TiKV encryption has some drawbacks that we are working actively to address in future versions.
```

> * When deploying a TiDB cluster, the majority of user data are stored in TiKV nodes, and those data are encrypted-at-rest when encryption is enabled. However there are small amount of user data stored in PD nodes as metadata. As of v4.0.0, PD doesn't support encryption-at-rest. As a workaround, it is recommended to use storage level encryption (e.g. file system encryption) to protect sensitive data stored in PD.

Reviewer: ```suggestion
* When deploying a TiDB cluster, the majority of a user's data is stored in TiKV nodes, and that data will be encrypted when encryption is enabled. However there is a small amount of user data stored in PD nodes as metadata. As of v4.0.0, PD doesn't support encryption-at-rest. We recommend using storage level encryption (e.g. file system encryption) to help protect sensitive data stored in PD.
```

Reviewer: More details on what is stored would be helpful here

> * TiKV currently does not exclude encryption keys and user data from core dump. It is highly adviced to disable core dump for TiKV process when using encryption-at-rest. This is not currently handled by TiKV itself.

Reviewer: ```suggestion
* TiKV currently does not exclude encryption keys and user data from core dumps. It is advised to disable core dumps for the TiKV process when using encryption-at-rest. This is not currently handled by TiKV itself.
```

> * TiKV keeps track of encryption key and method used for each data files, using absolute path of the files. As a result, once encryption is turned on for a TiKV node, user should not change data file paths config such as `storage.data-dir`, `raftstore.raftdb-path`, `rocksdb.wal-dir` and `raftdb.wal-dir`.

Reviewer: ```suggestion
* TiKV tracks encrypted data files using the absolute path of the files. As a result, once encryption is turned on for a TiKV node, the user should not change the data file paths config such as `storage.data-dir`, `raftstore.raftdb-path`, `rocksdb.wal-dir` and `raftdb.wal-dir`.
```

## BR S3 server-side encryption

To enable S3 server-side encryption when backup to S3 using BR, pass `--s3.sse` argument and set value to "aws:kms". S3 will use its own KMS key for encryption. Example:

```
./br backup full --pd <pd-address> --storage "s3://<bucket>/<prefix>" --s3.sse aws:kms
```

> To use a custom AWS KMS CMK that you created and owned, pass `--s3.sse-kms-key-id` in addition. In this case, both of the BR process and all the TiKV nodes in the cluster would need access to the KMS CMK, and the KMS CMK need to be in the same AWS region as the S3 bucket used for store the backup. It is adviced to grant access to the KMS CMK to BR process and TiKV nodes via AWS IAM. Please refer to AWS documentation for usage of IAM. Example:

Reviewer: ```suggestion
To use a custom AWS KMS CMK that you created and owned, pass `--s3.sse-kms-key-id` in addition. In this case, both the BR process and all the TiKV nodes in the cluster would need access to the KMS CMK (via AWS IAM), and the KMS CMK needs to be in the same AWS region as the S3 bucket used to store the backup. Please refer to AWS documentation for usage of IAM. Example:
```

```
./br backup full --pd <pd-address> --storage "s3://<bucket>/<prefix>" --s3.region <region> --s3.sse aws:kms --s3.sse-kms-key-id 0987dcba-09fe-87dc-65ba-ab0987654321
```

> On restore the backup, both `--s3.sse` and `--s3.sse-kms-key-id` should NOT be used, as S3 will figure out encryption settings by itself. The BR process and TiKV nodes in the cluster to restore the backup to would also need access to the KMS CMK, or the restore will fail. Example:

Reviewer: ```suggestion
When restoring the backup, both `--s3.sse` and `--s3.sse-kms-key-id` should NOT be used. S3 will figure out encryption settings by itself. The BR process and TiKV nodes in the cluster to restore the backup also need access to the KMS CMK. Example:
```

---
title: Encryption-At-Rest
summary: Learn how to enable encryption-at-rest to protect sensitive data.
category: reference
---

# Encryption-At-Rest for TiKV <span class="version-mark">New in v4.0.0</span>

> Encryption-at-rest means that data is encrypted when it is stored. For databases, this feature is also referred to as TDE (transparent data encryption). This is opposed to encryption in flight (TLS) or encryption in use (rarely used). Different things could be doing encryption-at-rest (SSD drive, file system, cloud vendor, etc), but by having TiKV do the encryption before storage this helps ensure that attackers must authenticate with the database to gain access to data. For example, when an attacker gains access to the physical machine, data cannot be accessed data by copying files on disk.

Reviewer: ```suggestion
Encryption-at-rest means that data is encrypted when it is stored. This is opposed to encryption in flight (TLS) or encryption in use (rarely used). For databases, encryption at rest is also referred to as TDE (transparent data encryption). Different layers can perform encryption-at-rest (SSD drive, file system, cloud vendor, etc), but by having TiKV do the encryption before storage this helps ensure that attackers must authenticate with the database to gain access to data. For example, when an attacker gains access to the physical machine, data cannot be accessed by copying files on disk.
```

> TiKV supports encryption-at-rest starting from v4.0.0. The feature allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation)  mode. To enable encryption-at-rest, an encryption key must be provided by user and this key is called master key. The master key can be provided via AWS KMS (recommended), or specifying a key stored as plaintext in a file. TiKV automatically rotate data keys that it used to encrypt actual data files. Manually rotating the master key can be done occassionally. Note that encryption-at-rest only encrypts data at rest (i.e. on disk) and not while data is transferred over network. It is advices to use TLS together with encryption-at-rest.

Reviewer: ```suggestion
TiKV supports encryption-at-rest starting from v4.0.0. The feature allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation)  mode. To enable encryption-at-rest, an encryption key must be provided by user and this key is called master key. The master key can be provided via AWS KMS (recommended), or specifying a key stored as plaintext in a file. TiKV automatically rotates data keys that it used to encrypt actual data files. Manually rotating the master key can be done occassionally. Note that encryption-at-rest only encrypts data at rest (i.e. on disk) and not while data is transferred over network. It is advices to use TLS together with encryption-at-rest.
```

Also from v4.0.0, BR supports S3 server-side encryption (SSE) when backing up to S3. A customer owned AWS KMS key can also be used together with S3 server-side encrytion.

## Caveats

The current version of TiKV encryption has some drawbacks that we are working actively to address in the future versions.

> * When deploying a TiDB cluster, the majority of a user's data is stored in TiKV nodes, and that data will be encrypted when encryption is enabled. However there is a small amount of user data stored in PD nodes as metadata (e.g secondary index keys used as TiKV region boundaries). As of v4.0.0, PD doesn't support encryption-at-rest. We recommend using storage level encryption (e.g. file system encryption) to help protect sensitive data stored in PD.

Reviewer: Is this the only source of data leakage? Or it is the biggest one?

Author: We currently believe the secondary index is the only case, but not confirmed.

> * TiKV currently does not exclude encryption keys and user data from core dumps. It is adviced to disable core dumps for the TiKV process when using encryption-at-rest. This is not currently handled by TiKV itself.

Reviewer: ```suggestion
* TiKV currently does not exclude encryption keys and user data from core dumps. It is advised to disable core dumps for the TiKV process when using encryption-at-rest. This is not currently handled by TiKV itself.
```

* TiKV tracks encrypted data files using the absolute path of the files. As a result, once encryption is turned on for a TiKV node, the user should not change data file paths config such as `storage.data-dir`, `raftstore.raftdb-path`, `rocksdb.wal-dir` and `raftdb.wal-dir`.

> * TiKV info log may contain user data for debugging purpose. Those data is not encrypted.

Reviewer: ```suggestion
* The TiKV info log contains user data for debugging purposes. The info log and this data in it are not encrypted.
```

TiKV supports encryption-at-rest starting from v4.0.0. The feature allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation)  mode. To enable encryption-at-rest, an encryption key must be provided by user and this key is called master key. The suggested way to provide the master key is via AWS KMS, but specifying a key in plaintext stored in a file is also supported. TiKV automatically rotate data keys that it used to encrypt actual data files. Manually rotating the master key can be done occassionally. Note that encryption-at-rest only encrypts data at rest (i.e. on disk) and not while data is transferred over network. It is advices to use TLS together with encryption-at-rest.

Also from v4.0.0, BR supports S3 server-side encryption (SSE) when backup to S3. A customer owned AWS KMS key can also be used together with S3 server-side encrytion.

> ## TiKV encryption-at-rest

Author: TODO: add a section on grafana monitoring for encryption
